### PR TITLE
Storage: Fix optimized BTRFS refresh

### PR DESCRIPTION
**Pull request by monstermunchkin**
Fixes https://github.com/lxc/lxd/issues/11308

### CODE REVIEW

File: lxd/storage/backend_lxd.go [hunk 1]
```diff
@@ -1387,7 +1387,7 @@ func (b *lxdBackend) RefreshCustomVolume(projectName string, srcProjectName stri
 // RefreshInstance synchronises one instance's volume (and optionally snapshots) over another.
 // Snapshots that are not present in the source but are in the destination are removed from the
 // destination if snapshots are included in the synchronisation.
-func (b *lxdBackend) RefreshInstance(inst instance.Instance, src instance.Instance, srcSnapshots []instance.Instance, allowInconsistent bool, op *operations.Operation) error {
+func (b *lxdBackend) RefreshInstance(inst instance.Instance, src instance.Instance, srcSnapshots []instance.Instance, allowInconsistent bool, snapshots bool, op *operations.Operation) error {
```
Comment (tomponline) [inline lxd/storage/backend_lxd.go:1390]: Can we use `srcSnapshots` argument rather than needing a new argument?

Comment (tomponline) [reply lxd/storage/backend_lxd.go:1390]: Yeah looking at where `RefreshInstance` is called, `srcSnapshots` should be nil in instance only mode. Whereas if there are no snapshots to sync it should be an empty slice I think.

**Review by tomponline — CHANGES_REQUESTED** [review]
Looks good, just a question around the need for an additional function argument.

**Comment by tomponline** [comment]
@ru-fu was there any documentation changes we need to make as part of these changes to how btrfs refresh works?

**Comment by ru-fu** [comment]
> @ru-fu was there any documentation changes we need to make as part of these changes to how btrfs refresh works?

No, I think this was just a bug fix to make btrfs work as intended. The relevant section is here and should be correct as far as I can tell:
https://linuxcontainers.org/lxd/docs/latest/reference/storage_drivers/#optimized-volume-transfer

We'll need a doc update when we switch to using rsync when there are no snapshots.

**Comment by tomponline** [comment]
> > @ru-fu was there any documentation changes we need to make as part of these changes to how btrfs refresh works?
> 
> No, I think this was just a bug fix to make btrfs work as intended. The relevant section is here and should be correct as far as I can tell: https://linuxcontainers.org/lxd/docs/latest/reference/storage_drivers/#optimized-volume-transfer
> 
> We'll need a doc update when we switch to using rsync when there are no snapshots.

Ah yeah, that is what it was now, thanks.
In that case we do need that for this PR as this commit includes that change https://github.com/lxc/lxd/pull/11337/commits/5e1661066c092bfb945cce7ede4c0bb9052de5e3

**Comment by ru-fu** [comment]
> Ah yeah, that is what it was now, thanks. In that case we do need that for this PR as this commit includes that change [5e16610](https://github.com/lxc/lxd/commit/5e1661066c092bfb945cce7ede4c0bb9052de5e3)

Is that for Btrfs only? Or for all storage drivers that have optimized transfer?

**Comment by tomponline** [comment]
> > Ah yeah, that is what it was now, thanks. In that case we do need that for this PR as this commit includes that change [5e16610](https://github.com/lxc/lxd/commit/5e1661066c092bfb945cce7ede4c0bb9052de5e3)
> 
> Is that for Btrfs only? Or for all storage drivers that have optimized transfer?

Thats a good point, we need something similar for ZFS and Ceph @monstermunchkin can you add that to this PR please?

As an aside, when you added optimized refresh, did that include Ceph (does it even support such a concept)?

@ru-fu BTRFS, Ceph and ZFS are the only ones that have an optimized transfer mechanism so they are the only ones that need to fallback to rsync for container-only migrations.

**Comment by ru-fu** [comment]
Here's my suggestion for the doc update.

Second paragraph:

```
LXD uses this optimized transfer when transferring instances and snapshots between storage pools that use the same storage driver, if the storage driver supports optimized transfer and the optimized transfer is actually quicker.
Otherwise, ...
```

Last paragraph:

```
On the other hand, refreshing copies of instances without snapshots (either because the instance doesn't have any snapshots or because the refresh uses the `--instance-only` flag) would actually be slower than using `rsync`.
In such cases, the optimized transfer would transfer the difference between the (non-existent) latest snapshot and the main volume, thus the full volume.
Therefore, LXD uses `rsync` instead of the optimized transfer for refreshes without snapshots.
```

**Comment by monstermunchkin** [comment]
> Thats a good point, we need something similar for ZFS and Ceph @monstermunchkin can you add that to this PR please?

> you might want to separate this into two PRs, one to fix the BTRFS optimized refresh, without the fallback to rsync, and the other adding fallback to rsync support for all optimized storage drivers that need it, along with the documentation.

Done and done: https://github.com/lxc/lxd/pull/11368

> As an aside, when you added optimized refresh, did that include Ceph (does it even support such a concept)?

@tomponline No, only btrfs and zfs IIRC.

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
@monstermunchkin out of interest, whats the storage around optimized refresh for ceph?

**Comment by monstermunchkin** [comment]
> @monstermunchkin out of interest, whats the storage around optimized refresh for ceph?

There's no optimized refresh for ceph. It always always calls `genericVFSCopyVolume` in `RefreshVolume`. I don't know why it wasn't added as it should be possible in theory.

**Pull request merged**